floors/floors.go

```diff
@@ -136,24 +138,35 @@ func isPriceFloorsEnabledForRequest(bidRequestWrapper *openrtb_ext.RequestWrappe
 	return true
 }
 
+// useFetchedData will check if to use fetched data or request data
+func useFetchedData(rate *int) bool {
+	if rate == nil {
+		return true
+	}
+	randomNumber := rand.Intn(dataRateMax)
+	return randomNumber < *rate
+}
+
 // resolveFloors does selection of floors fields from request data and dynamic fetched data if dynamic fetch is enabled
 func resolveFloors(account config.Account, bidRequestWrapper *openrtb_ext.RequestWrapper, conversions currency.Conversions, priceFloorFetcher FloorFetcher) (*openrtb_ext.PriceFloorRules, []error) {
-	var errList []error
-	var floorRules *openrtb_ext.PriceFloorRules
+	var (
+		errList     []error
+		floorRules  *openrtb_ext.PriceFloorRules
+		fetchResult *openrtb_ext.PriceFloorRules
+		fetchStatus string
+	)
 
 	reqFloor := extractFloorsFromRequest(bidRequestWrapper)
 	if reqFloor != nil && reqFloor.Location != nil && len(reqFloor.Location.URL) > 0 {
 		account.PriceFloors.Fetcher.URL = reqFloor.Location.URL
 	}
 	account.PriceFloors.Fetcher.AccountID = account.ID
 
-	var fetchResult *openrtb_ext.PriceFloorRules
-	var fetchStatus string
 	if priceFloorFetcher != nil && account.PriceFloors.UseDynamicData {
 		fetchResult, fetchStatus = priceFloorFetcher.Fetch(account.PriceFloors)
 	}
 
-	if fetchResult != nil && fetchStatus == openrtb_ext.FetchSuccess {
+	if fetchResult != nil && fetchStatus == openrtb_ext.FetchSuccess && useFetchedData(fetchResult.Data.FetchRate) {
```

Comment:
Nitpick: Although the `Fetch(config config.AccountPriceFloors)` implementation called in line 166 returns `openrtb_ext.FetchNone` when `fetchResult.Data` is `nil`, making it impossible for a `nil` `fetchResult.Data` in the third conditional in line 169, can we still add a check before dereferencing?
```
166         fetchResult, fetchStatus = priceFloorFetcher.Fetch(account.PriceFloors)
167     }
168
169 -   if fetchResult != nil && fetchStatus == openrtb_ext.FetchSuccess && useFetchedData(fetchResult.Data.FetchRate) {
    +   if fetchResult != nil && fetchStatus == openrtb_ext.FetchSuccess && fetchResult.Data != nil && useFetchedData(fetchResult.Data.FetchRate) {
170         mergedFloor := mergeFloors(reqFloor, fetchResult, conversions)
171         floorRules, errList = createFloorsFrom(mergedFloor, account, fetchStatus, openrtb_ext.FetchLocation)
172     } else if reqFloor != nil {
floors/floors.go
```
I guess to prevent panics in case the code changes in the future. Thoughts?

Reply:
In case of fetchStatus == openrtb_ext.FetchSuccess, fetchResult.Data can not be nil [https://github.com/prebid/prebid-server/blob/498955bcd5c242866a9b05886ac4e703cc850b2f/floors/fetcher.go#L140], so no need to add redundant check.